src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -50,151 +54,261 @@
-      COORD and ZCORN, or the various rectangular keywords like DX,DY
-      and DZ.
-
-      Actnum is treated specially:
-
-        1. If an actnum pointer is passed in that should be a pointer
-           to 0 and 1 values which will be used as actnum mask.
-
-        2. If the actnum pointer is not present the constructor will
-           look in the deck for an actnum keyword, and use that if it
-           is found. This is a best effort which will work in many
-           cases, but if the ACTNUM keyword is manipulated in the deck
-           those manipulations will be silently lost; if the ACTNUM
-           keyword has size different from nx*ny*nz it will also be
-           silently ignored.
-
-      With a mutable EclipseGrid instance you can later call the
-      EclipseGrid::resetACTNUM() method when you have complete actnum
-      information. The EclipseState based construction of EclipseGrid
-      is a two-pass operation, which guarantees that actnum is handled
-      correctly.
-    */
 
+EclipseGrid::EclipseGrid(const Deck& deck, const int * actnum)
+    : GridDims(deck),
+      m_minpvMode(MinpvMode::ModeEnum::Inactive),
+      m_pinch("PINCH"),
+      m_pinchoutMode(PinchMode::ModeEnum::TOPBOT),
+      m_multzMode(PinchMode::ModeEnum::TOP)
+{
+
+    if (deck.hasKeyword("GDFILE")){
+        
+        if (deck.hasKeyword("COORD")){
+            throw std::invalid_argument("COORD can't be used together with GDFILE");
+        }
+
+        if (deck.hasKeyword("ZCORN")){
+            throw std::invalid_argument("ZCORN can't be used together with GDFILE");
+        }
+        
+        if (deck.hasKeyword("ACTNUM")){ 
+            if (keywInputBeforeGdfile(deck, "ACTNUM"))  {
+                m_useActnumFromGdfile = true;
+            }
+        } else {
+            m_useActnumFromGdfile = true;
+        }
 
+    }
+        
+
+    const std::array<int, 3> dims = getNXYZ();
+
+    initGrid(dims, deck);
+       
+    if (deck.hasKeyword("MAPUNITS")){
+       if ((m_mapunits =="") || ( !keywInputBeforeGdfile(deck, "MAPUNITS"))) {
+            const auto& record = deck.getKeyword<ParserKeywords::MAPUNITS>( ).getStringData();
+            m_mapunits = record[0];
+       } 
+    }
+
+    if (deck.hasKeyword("MAPAXES")){
+       if ((m_mapaxes.size() == 0) || ( !keywInputBeforeGdfile(deck, "MAPAXES"))) {
+          const Opm::DeckKeyword& mapaxesKeyword = deck.getKeyword<ParserKeywords::MAPAXES>();
+          m_mapaxes.resize(6);
+          for (size_t n=0; n < 6; n++){
+             m_mapaxes[n] = mapaxesKeyword.getRecord(0).getItem(n).get<double>(0);
+          }
+       } 
+    }
+    
+    if (actnum != nullptr) {
+        
+        size_t nCells = dims[0]*dims[1]*dims[2];
+        
+        std::vector<int> actnumVector;
+        actnumVector.resize(nCells);
+        
+        for (size_t n = 0; n < nCells; n++){
+            actnumVector[n] = actnum[n];
+        }
```

Comment:
I guess `std::copy(actnumVector.begin(), actnumVector.end(), actnum);` would work - and earn you a deposit in the `std::` bank

Reply:
I see your suggestion and raise you one...
```
std::vector<int> actnumVector(actnum, actnum + nCells);
```

Reply:
fixed